### PR TITLE
Dehydrated devices

### PR DESCRIPTION
**Pull request by uhoreg**
Implements an alternate API for dehydration (compare with https://github.com/matrix-org/synapse/pull/7955)

Still needs unit tests, and is completely untested, but should be good enough to compare with 7955.

The idea is that the user first logs in as normal.  They then call `GET unstable/org.matrix.msc2697.v2/restore_device` to get the dehydrated device data.  If they are able to dehydrate the device, they call `POST unstable/org.matrix.msc2697.v2/restore_device` to claim the dehydrated device, and the server tells them if they can use it or not.

### CODE REVIEW

**Comment by clokep** [comment]
@uhoreg Thanks for taking a look at this!

My inclination is that this is simpler than #7955, but there could be gotchas I'm not seeing. It looks like much of the code here is identical to #7955, but we need to keep track of a bit less state. That seems like a good thing to me.

What are your thoughts on it? Do you see issues with this approach?

I didn't leave comments on any particular bits of code since I assumed you were looking for an overall approach. Let me know if you want a more thorough review!

**Comment by uhoreg** [comment]
@clokep  Thanks for the quick review!

> What are your thoughts on it? Do you see issues with this approach?

It definitely seems simpler.  The only issues I can see are the ones that I've already expressed (synapse having some hidden assumption somewhere that device IDs don't change, and that we're handing an access token to clients and expecting them not to mix dehydration calls in with other calls), but the first concern is probably just paranoia, and the second concern probably isn't too bad.

**Comment by clokep** [comment]
Removing from my review queue since I believe I've provided feedback. Let me know if I'm mistaken!

File: synapse/rest/client/v2_alpha/keys.py
Comment (clokep) [inline synapse/rest/client/v2_alpha/keys.py:81]: This comment needs to be updated to take into account the dehydrated device.

**Review by clokep — CHANGES_REQUESTED** [review]
Looks pretty good overall! I think most of my comments are straightforward, but let me know if there's questions!

File: synapse/storage/databases/main/devices.py
Comment (uhoreg) [reply synapse/storage/databases/main/devices.py:734]: Yeah, I was thinking that we already know if we should be doing an insert or an update, but I guess an upsert is clearer

File: synapse/rest/client/v2_alpha/devices.py
Comment (uhoreg) [reply synapse/rest/client/v2_alpha/devices.py:200]: > (And why does POST not allow guests?)

I probably copy-and-pasted from a different spot.  But you're right, it probably doesn't make sense for guest accounts.

File: synapse/rest/client/v2_alpha/devices.py
Comment (clokep) [inline synapse/rest/client/v2_alpha/devices.py:272]: I think this is missing a `not`:

```suggestion
        elif not isinstance(submission["device_id"], str):
```

File: synapse/rest/client/v2_alpha/devices.py
Comment (clokep) [inline synapse/rest/client/v2_alpha/devices.py:219]: I think this is missing a `not`:

```suggestion
        elif not isinstance(submission["device_id"], dict):
```

File: synapse/rest/client/v2_alpha/keys.py
Comment (clokep) [inline synapse/rest/client/v2_alpha/keys.py:82]:
```suggestion
            # Providing the device_id should only be done for setting keys
            # for dehydrated devices; however, we allow it for any device for
            # compatibility with older clients.
```

File: tests/handlers/test_device.py
Comment (clokep) [inline tests/handlers/test_device.py:303]:
```suggestion
        self.assertIsNone(ret)
```

**Review by clokep — CHANGES_REQUESTED** [review]
I think there's a couple of bugs in the REST layer, but overall looks good.

**Comment by clokep** [comment]
Note that this implements https://github.com/matrix-org/matrix-doc/pull/2697

**Comment by uhoreg** [comment]
One more change before this get merged: I want to make it copy the device display name from the current device to the dehydrated device when you rehydrate.

File: synapse/rest/client/v2_alpha/devices.py
Comment (clokep) [inline synapse/rest/client/v2_alpha/devices.py:194]:
```suggestion
        super().__init__()
```

File: synapse/rest/client/v2_alpha/devices.py
Comment (clokep) [inline synapse/rest/client/v2_alpha/devices.py:258]:
```suggestion
        super().__init__()
```

File: synapse/handlers/device.py
Comment (clokep) [inline synapse/handlers/device.py:578]: Can you add a comment above this saying why we aren't calling `self.delete_device` here? I *think* it is because we want to avoid the call to `delete_access_tokens_for_user`, but I'm not 100% sure.

File: synapse/handlers/device.py
Comment (clokep) [inline synapse/handlers/device.py:586]: It is probably clearer to fast-fail this method (`if not success: raise NotFoundError()`).

I also wonder if this method shouldn't return anything and the rest layer should generate the `{success: True}`? I think we tend to do that both ways though so I'll leave it up to you.

**Review by clokep — COMMENTED** [review]
Just a couple final nits! 👍

**Review by clokep — APPROVED** [review]
Seems great. Were we going to add something to the list of experimental features or is that not necessary?

**Comment by uhoreg** [comment]
I don't think it's necessary.  Clients can get by with just trying the endpoints and bailing if the endpoint doesn't exist.

**Comment by uhoreg** [comment]
Why does buildkite/synapse/sytest-python-3-dot-5-slash-postgres-9-dot-6-slash-workers keep failing?  It's failing on a test that is unrelated to my change, and all the other runners succeeded.  But I've re-run the test a few times and it's still failing. :-/

**Comment by clokep** [comment]
> Why does buildkite/synapse/sytest-python-3-dot-5-slash-postgres-9-dot-6-slash-workers keep failing? It's failing on a test that is unrelated to my change, and all the other runners succeeded. But I've re-run the test a few times and it's still failing. :-/

See matrix-org/sytest#969.

I just kicked it off again though so it should be successful this time.

**Pull request merged**
